core/src/test/java/org/infinispan/topology/ClusterTopologyManagerImplTest.java

```diff
@@ -0,0 +1,294 @@
+
+      MockLocalTopologyManager ltm = new MockLocalTopologyManager(CACHE_NAME);
+      gcr.registerComponent(ltm, LocalTopologyManager.class);
+
+      // Initial conditions (rebalance in phase 3, READ_NEW_WRITE_ALL)
+      transport.init(B, 2, asList(A, B));
+      ConsistentHash stableCH = chf.create(MurmurHash3.getInstance(), joinInfoA.getNumOwners(),
+                                           joinInfoA.getNumSegments(), singletonList(A), null);
+      ConsistentHash pendingCH = chf.create(MurmurHash3.getInstance(), joinInfoA.getNumOwners(),
+                                            joinInfoA.getNumSegments(), asList(A, B), null);
+      CacheTopology initialTopology = new CacheTopology(4, 2, stableCH, pendingCH,
+                                                        CacheTopology.Phase.READ_NEW_WRITE_ALL, asList(A, B),
+                                                        asList(joinInfoA.getPersistentUUID(),
+                                                               joinInfoB.getPersistentUUID()));
+      CacheTopology stableTopology = new CacheTopology(1, 1, stableCH, null,
+                                                       CacheTopology.Phase.NO_REBALANCE, singletonList(A),
+                                                       singletonList(joinInfoA.getPersistentUUID()));
+      ltm.init(joinInfoA, initialTopology, stableTopology, AvailabilityMode.AVAILABLE);
+      // Normally LocalTopologyManagerImpl.start()/doHandleTopologyUpdate() registers the persistent UUIDs
+      // TODO Write test with asymmetric caches leaving the PersistentUUIDManager cache incomplete
+      persistentUUIDManager.addPersistentAddressMapping(A, joinInfoA.getPersistentUUID());
+      persistentUUIDManager.addPersistentAddressMapping(B, joinInfoB.getPersistentUUID());
+
+      // Component under test: ClusterTopologyManagerImpl on the new coordinator (B)
+      ClusterTopologyManagerImpl ctm = new ClusterTopologyManagerImpl();
+      gcr.registerComponent(ctm, ClusterTopologyManager.class);
+
+      // As CTMI starts as regular member it will request the rebalancing status from the coordinator
+      transport.expectTopologyCommand(CacheTopologyControlCommand.Type.POLICY_GET_STATUS)
+               .singleResponse(A, SuccessfulResponse.create(true));
+
+      ctm.start();
+      transport.verifyTopologyCommand(CacheTopologyControlCommand.Type.POLICY_GET_STATUS).assertDone();
+
+      // The coordinator (node A) leaves the cluster
+      transport.updateView(3, singletonList(B));
+      managerNotifier.notifyViewChange(singletonList(B), asList(A, B), B, 3);
+
+      // Node B becomes coordinator and CTMI tries to recover the cluster status
+      transport.verifyTopologyCommand(CacheTopologyControlCommand.Type.GET_STATUS).finish();
+
+      // CTMI broadcasts the only cache topology it got, but without the pending CH
+      ltm.expectTopology(5, singletonList(A), null, CacheTopology.Phase.NO_REBALANCE);
```

Comment:
is this meant to be `singletonList(B)`? The test fails otherwise

Reply:
This line is correct, the first topology after the merge (5) is with `A` as the only member. The failure was actually caused by a problem in `MockLocalTopologyManager.verifyTopology()`: it was re-reading the `status` field instead of using the `topology` parameter, and it was seeing the next topology (6), which is broadcast immediately after with `B` as the only member. In my defense, it never failed with trace logging enabled ;)

After your PR is in we need to simplify this and skip the useless topology 5, but I feel the need to add more tests before making that change.

Reply:
I think I'm missing something. Why is topology 5 useless? Is this not needed to "Increment the topology id so that it's bigger than any topology that might have been sent by the old coordinator."

Regarding my PR, feel free to review/merge :stuck_out_tongue_winking_eye:

Reply:
I meant we could install topology 6 directly (but with topology id 5). We still have to install a topology with a higher id after recovery, but there's no point in installing a topology that we know is going to be immediately replaced. In fact, since topology updates are not ordered, it's possible for some nodes to process topology 6 first and to discard topology 5 when they receive it.

When I first wrote `PreferAvailabilityStrategy` I was thinking that it shouldn't worry about removing members and reassigning segments to keep it simpler, but with all the recent additions it's probably a good time to rethink how responsibilities are split between the `AvailabilityStrategy` impl and `ClusterCacheStatus`.

Reply:
@danberindei Ok thanks for the clarification. I think your right about rethinking how responsibilities are split, it would be a massive win if we can reduce the complexity of some of the code :+1: